src/domain/substitution.ts

```diff
@@ -0,0 +1,30 @@
+export type SubstitutableVar = 'OWNER' | 'REPO' | 'TAG' | 'VERSION';
+
+export const ALL_SUBSTITUTABLE_VARS: SubstitutableVar[] = [
+  'OWNER',
+  'REPO',
+  'TAG',
+  'VERSION',
+];
+
+export function substituteVars(
+  str: string,
+  vars: Partial<Record<SubstitutableVar, string>>
+): string {
+  for (const key of Object.keys(vars)) {
+    str = str.replaceAll(`{${key}}`, vars[key as SubstitutableVar]);
+  }
+  return str;
+}
+
+export function getUnsubstitutedVars(str: string): Set<SubstitutableVar> {
```

Comment:
I thought unsubstituted would be whatever is leftover after substituting the known keys. Maybe it should have a different name?

Reply:
That is the meaning---unless our wires are crossed. This checks for any known `{VAR}`s in the string that have not been substituted with a value.

Reply:
Not substituted yet, simply because the substitution hasn't run yet right? Not that they're unknown VAR that failed to be substituted. I'm probably just expecting a residual for some reason, not important.

Reply:
The former, either because substitution has not been run, or it has been run but not with all the vars in the string. For example, in the CLI you need to pass the `--github-repository` in order to substitute `{OWNER}` and `{REPO}` because unlike the app that information isn't readily available.

This is used in `SourceTemplate::validateFullySubstituted()` and the equivalent in the new attestations template.

Variables that are unrecognized are just ignored and considered to be part of the string.

Reply:
Will land this but can re-visit if you still feel the name is off.